text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+- Framework versions might require an upgrade of the CLI. In which case, we provide instructions as to which CLI version should be used.
+
+This RFC details the necessary strategy and mechanism for ensuring that we keep those promises when releasing new versions.
+
+# Motivation
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
+- An incompatible change in the framework, will always be accompanied with a protocol version bump.
+- An incompatible change in the framework, accompanied with a protocol version bump, will always produce the correct error message.
+
+There are 2 main motivations why we should put some substantial effort into this:
+
+## (1) Haven't we suffered enough?
+
+Recently, we experienced several incidents that are directly related to this domain:
+
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.21.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/5986)
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.16.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/4897)
+- [aws/aws-cdk: CDK CLI can only be used with apps created by CDK >= 1.10.0](https://github.com/aws/aws-cdk/issues/4294)
+
+*We want to implement this RFC to prevent (or at least greatly reduce) the occurrences of such issues.*
+
+## (2) Refactor CX Protocol versioning
+
+We've talked about the mechanism we currently use to perform protocol version changes quite a bit. The process and the expected output aren't very clear, and we seem to be always trying to understand it again and again.
+This might warrant a re-think/re-design, to make it much more ergonomic and developer friendly. 
+
+However, performing changes on such a crucial part of the system, shouldn't be done without a proper validation mechanism.
+
+*We want to implement this RFC so that we are able to iterate and change the versioning mechanism itself* 
+
+Actually, We have already acknowledged the need for this a while back:
+
+- [aws/aws-cdk: cx-protocol backwards compatibility integration tests](https://github.com/aws/cdk-ops/issues/365)
+- [aws/aws-cdk: Design: Strategy for automated backwards-compatibility testing](https://github.com/aws/cdk-ops/issues/233)
+
+# Design Summary
+
+In order to ensure compatibility between the CLI and the Framework, we have to actually run the CLI against an old framework version, and vice versa.
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
```

Comment:
This statement is a bit confusing and feels like it's contradicting the following paragraph that says that we need to use old test suite when a new CLI runs against an old framework.

Reply:
Your'e right, this of course also refers to old integration tests. I think removing the word `current` will suffice.